### PR TITLE
Add last sequence number map to BlockMessage 

### PR DESCRIPTION
**Pull request by KentShikama**
## Overview
<sup>_What this PR does, and why it's needed_</sup>

This is information that the constant time estimator will need. See https://ethresear.ch/t/another-attempt-at-o-n-lmd-ghost-that-is-constant-time-on-the-number-of-blocks-since-genesis/4949 for constant time estimator specification.

### JIRA ticket:
<sup>_Create it if there isn't one already._</sup>

This is part of https://rchain.atlassian.net/browse/RCHAIN-2933

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by KentShikama** [comment]
bors try

**Comment by KentShikama** [comment]
bors try

File: casper/src/main/scala/coop/rchain/casper/Validate.scala [hunk 1]
```diff
@@ -418,6 +422,43 @@ object Validate {
       } yield Left(InvalidDeployCount)
     }
 
+  // TODO: Double check logic with validator rotation
+  def lastSequenceNumber[F[_]: Monad: Log: BlockStore](
```
Comment (odeac) [inline casper/src/main/scala/coop/rchain/casper/Validate.scala]: The name of this fucntion is misleading. Its name implies a computed lastSEquenceNumber while it actually does a validation

File: casper/src/main/scala/coop/rchain/casper/Validate.scala [hunk 2]
```diff
@@ -418,6 +422,43 @@ object Validate {
...
+                     Applicative[F].pure(Right(Valid))
+                   } else {
+                     val logMessage =
+                       s"main parent is missing even though block is not genesis."
+                     for {
+                       _ <- Log[F].warn(ignore(b, logMessage))
+                     } yield Left(InvalidLastSequenceNumber)
+                   }
+               }
+    } yield result
```
Comment (odeac) [inline casper/src/main/scala/coop/rchain/casper/Validate.scala]: I find this function difficult to read. can you make it more readable?

File: casper/src/main/scala/coop/rchain/casper/Validate.scala (hunk 1 above)
Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/Validate.scala]: Except it is in the Validate class. What do you suggest?

File: casper/src/main/scala/coop/rchain/casper/Validate.scala (hunk 2 above)
Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/Validate.scala]: Hmm should I extract the two cases?

File: casper/src/main/scala/coop/rchain/casper/Validate.scala (hunk 1 above)
Comment (odeac) [reply casper/src/main/scala/coop/rchain/casper/Validate.scala]: Actually you are right. I didn't get that the class name is the verb here. Never mind this comment

File: casper/src/main/scala/coop/rchain/casper/Validate.scala (hunk 2 above)
Comment (dzajkowski) [reply casper/src/main/scala/coop/rchain/casper/Validate.scala]: maybe:
```
  // TODO: Double check logic with validator rotation
  def lastSequenceNumber[F[_]: Monad: Log: BlockStore](
    b: BlockMessage,
    genesis: BlockMessage
  ): F[Either[InvalidBlock, ValidBlock]] = {
    def valid = (Valid:ValidBlock).asRight[InvalidBlock].pure[F]
    def invalid(logMessage: String) = Log[F].warn(ignore(b, logMessage)).as((InvalidLastSequenceNumber: InvalidBlock).asLeft[ValidBlock])
    def calculateLastSequenceNumber(mainParent: BlockMessage) = {
      val mainParentLastSequenceNumberMap =
        ProtoUtil.toLastSequenceNumberMap(mainParent.lastSequenceNumber)
      val computedLastSequenceNumber =
        mainParentLastSequenceNumberMap.updated(b.sender, b.seqNum)
      ProtoUtil.toLastSequenceNumber(computedLastSequenceNumber)
    }

    if (b == genesis)
      valid
    else
      ProtoUtil.mainParent[F](b) >>= {
        case Some(mainParent) =>
          if (b.lastSequenceNumber == calculateLastSequenceNumber(mainParent))
            valid
          else
            invalid("computed last sequence number is not equal to given last sequence number.")

        case None =>
          invalid("main parent is missing even though block is not genesis.")
      }
  }
```

Comment (odeac) [reply casper/src/main/scala/coop/rchain/casper/Validate.scala]: I'm thinking about something like this:
```scala
  def lastSequenceNumber[F[_]: Monad: Log: BlockStore](
      b: BlockMessage,
      genesis: BlockMessage
  ): F[Either[InvalidBlock, ValidBlock]] = {

    def computeLastSequenceNumber(mainParent : BlockMessage) = {
      val mainParentLastSequenceNumberMap =
        ProtoUtil.toLastSequenceNumberMap(mainParent.lastSequenceNumber)
      val computedLastSequenceNumberMap =
        mainParentLastSequenceNumberMap.updated(b.sender, b.seqNum)
      ProtoUtil.toLastSequenceNumber(computedLastSequenceNumberMap)
    }

    def computeErrors(maybeMainParentSequenceNumber : Option[Seq[LastSequenceNumber]]) =
      maybeMainParentSequenceNumber match {
         case Some(computedLastSequenceNumber) if (b.lastSequenceNumber != computedLastSequenceNumber) =>
           Some("computed last sequence number is not equal to given last sequence number.")
         case None if (b != genesis) => 
           Some("main parent is missing even though block is not genesis.")
         case _ => None
      }

    for {
      maybeParent <- ProtoUtil.mainParent[F](b)
    
      result <- computeErrors(maybeParent.map(computeLastSequenceNumber)) match {
        case None => Right(Valid).pure[F]
          
        case Some(errorMsg) => 
          Log[F].warn(ignore(b, errorMsg)) *> 
            Left(InvalidLastSequenceNumber).pure[F]
      }
    } yield result
  }```

File: casper/src/test/scala/coop/rchain/casper/HashSetCasperTest.scala [hunk 3]
```diff
@@ -1692,12 +1692,15 @@ class HashSetCasperTest extends FlatSpec with Matchers with Inspectors {
     val serializedBlockHash = ByteString.copyFrom(blockHash)
     val blockThatPointsToInvalidBlock =
       BlockMessage(serializedBlockHash, Some(header), Some(body), serializedJustifications)
-    nodes(1).casperEff.blockDag.flatMap { dag =>
-      ProtoUtil.signBlock[Effect](
+    val sender = ByteString.copyFrom(validators(1))
```
Comment (dzajkowski) [inline casper/src/test/scala/coop/rchain/casper/HashSetCasperTest.scala]: does it need to be a named val?

**Comment by dzajkowski** [comment]
bors try

**Comment by dzajkowski** [comment]
bors try

**Comment by dzajkowski** [comment]
bors try-

**Comment by dzajkowski** [comment]
bors try

**Comment by dzajkowski** [comment]
bors try

**Comment by dzajkowski** [comment]
bors try-

File: casper/src/main/scala/coop/rchain/casper/Validate.scala (hunk 2 above)
Comment (odeac) [reply casper/src/main/scala/coop/rchain/casper/Validate.scala]: ...or the for could be written like this:
```scala
    for {
      maybeParent <- ProtoUtil.mainParent[F](b)
      maybeLastSequenceNumber = maybeParent.map(computeLastSequenceNumber)
      result <-
        computeErrors(maybeLastSequenceNumber)
          .traverse(errorMsg => Log[F].warn(ignore(b, errorMsg)) *> Left(InvalidLastSequenceNumber).pure[F])
    } yield result.getOrElse(Right(Valid))
```

**Pull request closed without merging**